src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -44,6 +181,64 @@ private void SetContentType(ActionContext context, FileResult result)
             response.ContentType = result.ContentType;
         }
 
+        protected void ComputeIfMatch(ActionContext context, FileResult result, DateTimeOffset lastModified, EntityTagHeaderValue etag)
+        {
+            // 14.24 If-Match
+            var httpContext = context.HttpContext;
+            var httpRequestHeaders = httpContext.Request.GetTypedHeaders();
+            var ifMatch = httpRequestHeaders.IfMatch;
+            if (ifMatch != null && ifMatch.Any())
+            {
+                _ifMatchState = PreconditionState.PreconditionFailed;
+                foreach (var entityTag in ifMatch)
+                {
+                    if (etag.Equals(EntityTagHeaderValue.Any) || etag.Compare(etag, useStrongComparison: true))
+                    {
+                        _ifMatchState = PreconditionState.ShouldProcess;
+                        break;
+                    }
+                }
+            }
+
+            // 14.26 If-None-Match
+            var ifNoneMatch = httpRequestHeaders.IfNoneMatch;
+            if (ifNoneMatch != null && ifNoneMatch.Any())
+            {
+                _ifNoneMatchState = PreconditionState.ShouldProcess;
+                foreach (var entityTag in ifNoneMatch)
+                {
+                    if (etag.Equals(EntityTagHeaderValue.Any) || etag.Compare(etag, useStrongComparison: true))
+                    {
+                        _ifNoneMatchState = PreconditionState.NotModified;
+                        break;
+                    }
+                }
+            }
+        }
+
+        protected void ComputeIfModifiedSince(ActionContext context, DateTimeOffset lastModified)
+        {
+            var now = DateTimeOffset.UtcNow;
+            var httpContext = context.HttpContext;
+            var httpRequestHeaders = httpContext.Request.GetTypedHeaders();
+
+            // 14.25 If-Modified-Since
+            var ifModifiedSince = httpRequestHeaders.IfModifiedSince;
+            if (ifModifiedSince.HasValue && ifModifiedSince <= now)
+            {
+                bool modified = ifModifiedSince < lastModified;
+                _ifModifiedSinceState = modified ? PreconditionState.ShouldProcess : PreconditionState.NotModified;
+            }
+
+            // 14.28 If-Unmodified-Since
+            var ifUnmodifiedSince = httpRequestHeaders.IfUnmodifiedSince;
+            if (ifUnmodifiedSince.HasValue && ifUnmodifiedSince <= now)
+            {
+                bool unmodified = ifUnmodifiedSince >= lastModified;
+                _ifUnmodifiedSinceState = unmodified ? PreconditionState.ShouldProcess : PreconditionState.PreconditionFailed;
```

Comment:
These classes are singletons, you can't add fields to them as one instance is shared across all requests of the app